src/lib/idmap/sss_idmap.c

```diff
@@ -1078,6 +1170,180 @@ add_dom_for_sid(struct sss_idmap_ctx *ctx,
+        return IDMAP_OUT_OF_MEMORY;
+    }
+
+    ret = snprintf(out, len + 1, "%"PRIu32, id);
+    if (ret != len) {
+        ctx->free_func(out, ctx->alloc_pvt);
+        return IDMAP_ERROR;
+    }
+
+    *_out = out;
+    return IDMAP_SUCCESS;
+}
+
+struct sss_idmap_offset_murmurhash3_data offset_murmurhash3_data_default = { 0xdeadbeef };
+
+enum idmap_error_code sss_idmap_offset_murmurhash3(void *pvt,
+                                                   uint32_t range_size,
+                                                   const char *input,
+                                                   long long *offset)
+{
+    struct sss_idmap_offset_murmurhash3_data *offset_murmurhash3_data;
+    long long out;
+
+    if (input == NULL || offset == NULL) {
+        return IDMAP_ERROR;
+    }
+
+    if (pvt != NULL) {
+        offset_murmurhash3_data = (struct sss_idmap_offset_murmurhash3_data *) pvt;
+    } else {
+        offset_murmurhash3_data = &offset_murmurhash3_data_default;
+    }
+
+    out = murmurhash3(input, strlen(input), offset_murmurhash3_data->seed);
+
+    out %= range_size;
+
+    *offset = out;
+
+    return IDMAP_SUCCESS;
+}
+
+enum idmap_error_code sss_idmap_gen_to_unix(struct sss_idmap_ctx *ctx,
+                                            const char *domain_id,
+                                            const char *input,
+                                            uint32_t *_id)
+{
+    struct idmap_domain_info *idmap_domain_info;
+    struct idmap_domain_info *matched_dom = NULL;
+    long long offset;
+    uint32_t range_size;
+    enum idmap_error_code err;
+    idmap_offset_func *offset_func = sss_idmap_offset_murmurhash3;
+    void *offset_func_pvt = NULL;
+
+    if (domain_id == NULL || input == NULL || _id == NULL) {
+        return IDMAP_ERROR;
+    }
+
+    CHECK_IDMAP_CTX(ctx, IDMAP_CONTEXT_INVALID);
+
+    idmap_domain_info = ctx->idmap_domain_info;
+
+    if (idmap_domain_info->offset_func != NULL) {
+        offset_func = idmap_domain_info->offset_func;
+        if (idmap_domain_info->offset_func_pvt != NULL) {
+            offset_func_pvt = idmap_domain_info->offset_func_pvt;
+        }
+    }
+
+    /* Try primary slices */
+    while (idmap_domain_info != NULL) {
+
+        if (is_from_dom(idmap_domain_info->sid, domain_id)) {
+
+            if (idmap_domain_info->external_mapping == true) {
+                return IDMAP_EXTERNAL;
+            }
+
+            range_size = 1 + (idmap_domain_info->range_params.max_id - idmap_domain_info->range_params.min_id);
+            err = offset_func(offset_func_pvt, range_size, input, &offset);
+            if (err != IDMAP_SUCCESS) {
+                return err;
+            }
+
+            if (offset >= range_size) {
+                return IDMAP_ERROR;
+            }
+
+            if (comp_id(&idmap_domain_info->range_params, offset, _id)) {
+                return IDMAP_SUCCESS;
+            }
+
+            matched_dom = idmap_domain_info;
```

Comment:
I think we can just `return IDMAP_NO_RANGE` here?
I mean, we already found matching domain (but for some reason can't convert input to ID). There can't be other domain with the same `domain_id` in the context, right?

Reply:
No, in the IPA case, you can add additional id-ranges on the IPA server if the default one is not large enough and the additional one will use the same domain_id (domain SID of the trusted AD domain).